File: src/10-word-count/haskell/main.hs
{-
 * Polyglot Programming
 *
 * Copyright (c) 2024 Adam Burucs
 *
 * MIT Licensed
-}

import System.IO
import System.IO.Error (catchIOError)
import Control.Exception (try, SomeException)
import Data.Text (Text, splitOn, unpack)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO

countWords :: FilePath -> IO (Either String Int)
countWords fileName = do
    result <- try (TIO.readFile fileName) :: IO (Either SomeException Text)
    case result of
        Left ex -> return $ Left (show ex)
        Right content -> return $ Right (sum $ map (length . T.words) (T.lines content))

main :: IO ()
main = do
    putStrLn "Count total words sum"
    putStrLn ""

    countResult <- countWords "enchiridion-hun.txt"
    case countResult of
        Left err -> putStrLn $ "Error: " ++ err
        Right count -> putStrLn $ "Total words: " ++ show count

File: src/10-word-count/rust/src/main.rs
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024 Adam Burucs
 *
 * MIT Licensed
 */

use std::{
    fs::File,
    io::{BufRead, BufReader},
};

fn count_words(filepath: &str) -> Result<usize, Box<dyn std::error::Error>> {
    let mut count: usize = 0;
    let file = File::open(filepath)?;
    let reader = BufReader::new(file);

    for line in reader.lines() {
        match line {
            Ok(val) => {
                count += val.split_whitespace().count();
            }
            Err(err) => {
                println!("Error when matching lines {err}");
            }
        }
    }

    Ok(count)
}

fn main() {
    println!("Count total words sum");
    println!();
    // Example file
    let count = count_words("enchiridion-hun.txt");
    match count {
        Ok(val) => println!("Total word count: {val}"),
        Err(err) => println!("Error when counting words: {err}"),
    }
}
